### PR TITLE
adding tcp_info struct to linux musl/glibc.

### PR DESCRIPTION
**Pull request by devnexen**

### CODE REVIEW

**Review by JohnTitor — APPROVED** [review]

**Comment by JohnTitor** [comment]
Could you check the CI failure? Seems musl gets confused 🤔

**Comment by JohnTitor** [comment]
Could you fix the CI failure?
@rustbot author

**Comment by devnexen** [comment]
shame on me I did not realise the existence of ci shell scripts until now, pretty helpful.

**Comment by tgross35** [comment]
@devnexen I think this looks pretty good, could you rebase?

File: src/unix/linux_like/linux/musl/mod.rs [hunk 1]
```diff
@@ -344,6 +344,75 @@ s! {
...
+
+    pub struct tcp_info {
+        pub tcpi_state: u8,
+        pub tcpi_ca_state: u8,
+        pub tcpi_retransmits: u8,
+        pub tcpi_probes: u8,
+        pub tcpi_backoff: u8,
+        pub tcpi_options: u8,
+        /*
+         * FIXME: when musl headers are more up to date
```
Comment (tgross35) [inline src/unix/linux_like/linux/musl/mod.rs:358]:
```suggestion
         * FIXME(musl): when musl headers are more up to date
```

Just so we can grep these things easy at the musl update

File: src/unix/linux_like/linux/musl/mod.rs [hunk 2]
```diff
@@ -344,6 +344,75 @@ s! {
...
+        pub tcpi_busy_time: u64,
+        pub tcpi_rwnd_limited: u64,
+        pub tcpi_sndbuf_limited: u64,
+        pub tcpi_delivered: u32,
+        pub tcpi_delivered_ce: u32,
+        pub tcpi_bytes_sent: u64,
+        pub tcpi_bytes_retrans: u64,
+        pub tcpi_dsack_dups: u32,
+        pub tcpi_reord_seen: u32,
+        // FIXME: to uncomment once CI musl is updated
```
Comment (tgross35) [inline src/unix/linux_like/linux/musl/mod.rs:412]:
```suggestion
        // FIXME(musl): to uncomment once CI musl is updated
```

Same as above

File: src/unix/linux_like/linux/gnu/mod.rs [hunk 3]
```diff
@@ -453,6 +453,45 @@ s! {
...
+        pub tcpi_state: u8,
+        pub tcpi_ca_state: u8,
+        pub tcpi_retransmits: u8,
+        pub tcpi_probes: u8,
+        pub tcpi_backoff: u8,
+        pub tcpi_options: u8,
+        // bitfields 4
+        pub tcpi_snd_wscale: u8,
+        // bitfields 4
+        pub tcpi_rcv_wscale: u8,
```
Comment (tgross35) [inline src/unix/linux_like/linux/gnu/mod.rs:469]: The C header has these as 

```c
uint8_t	tcpi_snd_wscale : 4, tcpi_rcv_wscale : 4;
```

doesn't that mean `tcpi_snd_wscale` and `tcpi_rcv_wscale` are located within the same `u8`, and should be a single field in the Rust struct?

**Comment by tgross35** [comment]
Checking headers https://github.com/kraj/glibc/blob/1927f718fcc48bdaea03086bdc2adf11279d655b/sysdeps/gnu/netinet/tcp.h#L226-L267 and https://github.com/kraj/musl/blob/2c41ee96f60074fa8ac387e1fcdedba0654fb173/include/netinet/tcp.h#L195-L250 this looks good, with some notes above.

@rustbot author aince I think this might need a change for the bitfield

File: src/unix/linux_like/linux/gnu/mod.rs (hunk 3 above)
Comment (tgross35) [reply src/unix/linux_like/linux/gnu/mod.rs:469]: I was trying to find what we do here in other cases, but it looks like this doesn't come up very often. Probably just make it one field and add a comment so users know:

```rust
/// The bitfield containing `tcpi_snd_wscale` and `tcpi_rcv_wscale`, four bits each.
pub tcpi_snd_rcv_wscale: u8
```

The musl section that is commented out should probably say something similar.

File: src/unix/linux_like/linux/gnu/mod.rs [hunk 4]
```diff
@@ -453,6 +453,43 @@ s! {
...
+
+    pub struct tcp_info {
+        pub tcpi_state: u8,
+        pub tcpi_ca_state: u8,
+        pub tcpi_retransmits: u8,
+        pub tcpi_probes: u8,
+        pub tcpi_backoff: u8,
+        pub tcpi_options: u8,
+        // bitfields 4: tcp_snd_wscale, tcp_rcv_wscale
+        pub tcpi_snd_rcv_wscale: u8,
```
Comment (tgross35) [inline src/unix/linux_like/linux/gnu/mod.rs:467]:
```suggestion
        /// This contains the bitfields `tcpi_snd_wscale` and `tcpi_rcv_wscale`. Each is four bits.
        pub tcpi_snd_rcv_wscale: u8,
```

Let's make this show up in documentation

File: src/unix/linux_like/linux/musl/mod.rs [hunk 5]
```diff
@@ -344,6 +344,73 @@ s! {
...
+        pub tcpi_state: u8,
+        pub tcpi_ca_state: u8,
+        pub tcpi_retransmits: u8,
+        pub tcpi_probes: u8,
+        pub tcpi_backoff: u8,
+        pub tcpi_options: u8,
+        /*
+         * FIXME(musl): when musl headers are more up to date
+        // bitfields 4: tcp_snd_wscale, tcp_rcv_wscale
+        pub tcpi_snd_rcv_wscale: u8,
```
Comment (tgross35) [inline src/unix/linux_like/linux/musl/mod.rs:360]:
```suggestion
        /// This contains the bitfields `tcpi_snd_wscale` and `tcpi_rcv_wscale`. Each is four bits.
        pub tcpi_snd_rcv_wscale: u8,
```

Same as above

File: src/unix/linux_like/linux/musl/mod.rs [hunk 6]
```diff
@@ -344,6 +344,73 @@ s! {
...
+        // bitfields 1
+        pub tcpi_delivery_rate_app_limited: u8,
+        // bitfields 2
+        pub tcpi_fastopen_client_fail: u8,
```
Comment (tgross35) [inline src/unix/linux_like/linux/musl/mod.rs:364]: I think this needs the bitfield change too

```suggestion
        /// This contains the bitfields `tcpi_delivery_rate_app_limited` (one bit) and
        /// `tcpi_fastopen_client_fail` (two bits).
        pub tcpi_delivery_fastopen_bitfields: u8,
```

**Review by tgross35 — APPROVED** [review]

**Comment by tgross35** [comment]
Thanks!

@rustbot label +stable-nominated

**Pull request merged**